packages/language-server/src/plugins/typescript/service.ts

```diff
@@ -91,22 +101,107 @@ export async function getServiceForTsconfig(
     if (services.has(tsconfigPath)) {
         service = await services.get(tsconfigPath)!;
     } else {
-        Logger.log('Initialize new ts service at ', tsconfigPath);
+        const reloading = pendingReloads.has(tsconfigPath);
+
+        if (reloading) {
+            Logger.log('Reloading ts service at ', tsconfigPath, ' due to config updated');
+        } else {
+            Logger.log('Initialize new ts service at ', tsconfigPath);
+        }
+
+        pendingReloads.delete(tsconfigPath);
         const newService = createLanguageService(tsconfigPath, docContext);
         services.set(tsconfigPath, newService);
         service = await newService;
+
+        updateExtendedConfigDependents(service);
+        watchConfigFile(service, docContext);
     }
 
     return service;
 }
 
+function updateExtendedConfigDependents(service: LanguageServiceContainer) {
+    service.extendedConfigPaths.forEach((extendedConfig) => {
+        let dependedTsConfig = extendedConfigToTsConfigPath.get(extendedConfig);
+        if (!dependedTsConfig) {
+            dependedTsConfig = new Set();
+            extendedConfigToTsConfigPath.set(extendedConfig, dependedTsConfig);
+        }
+
+        dependedTsConfig.add(service.tsconfigPath);
+    });
+}
+
+function watchConfigFile(ls: LanguageServiceContainer, docContext: LanguageServiceDocumentContext) {
+    if (!ts.sys.watchFile || !docContext.watchTsConfig) {
+        return;
+    }
+
+    if (!configWatchers.has(ls.tsconfigPath) && ls.tsconfigPath) {
+        configWatchers.set(
+            ls.tsconfigPath,
+            ts.sys.watchFile(ls.tsconfigPath, (filename, kind) =>
+                watchConfigCallback(filename, kind, docContext)
+            )
+        );
+    }
+
+    for (const config of ls.extendedConfigPaths) {
+        if (extendedConfigWatchers.has(config)) {
+            continue;
+        }
+        extendedConfigWatchers.set(
+            config,
+            ts.sys.watchFile(config, (filename, kind) =>
+                watchExtendedConfigCallback(filename, kind, docContext)
+            )
+        );
+    }
+}
+
+async function watchExtendedConfigCallback(
+    extendedConfig: string,
+    kind: ts.FileWatcherEventKind,
+    docContext: LanguageServiceDocumentContext
+) {
+    docContext.extendedConfigCache.delete(extendedConfig);
+
+    extendedConfigToTsConfigPath.get(extendedConfig)?.forEach((config) => {
+        services.delete(config);
+        pendingReloads.add(config);
```

Comment:
I was at first confused why this is enough, until I realized that you don't need to explicitly restart the server because the next `getService` call will do this. Could you add a comment here and in `watchConfigCallback` about this?